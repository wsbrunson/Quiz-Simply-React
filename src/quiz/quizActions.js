// @flow
import type {
  TypeQuestionsMap,
  TypeSelectionsMap,
} from '../types/TypeQuiz.flow';

const ActionTypes = {
  REQUEST_QUIZ: 'REQUEST_QUIZ',
  REQUEST_QUIZ_SUCCEEDED: 'REQUEST_QUIZ_SUCCEEDED',
  REQUEST_QUIZ_FAILED: 'REQUEST_QUIZ_FAILED',
  SELECT_CHOICE: 'SELECT_CHOICE',
};

export type TypeRequestQuizAction = {
  type: typeof ActionTypes.REQUEST_QUIZ,
};

const requestQuiz = (): TypeRequestQuizAction => ({
  type: ActionTypes.REQUEST_QUIZ,
});

export type TypeRequestQuizSucceededAction = {
  type: typeof ActionTypes.REQUEST_QUIZ_SUCCEEDED,
  payload: {
    questions: TypeQuestionsMap,
  },
};

const requestQuizSucceeded = (
  questions: TypeQuestionsMap,
): TypeRequestQuizSucceededAction => ({
  type: ActionTypes.REQUEST_QUIZ_SUCCEEDED,
  payload: {
    questions,
  },
});

export type TypeRequestQuizFailedAction = {
  type: typeof ActionTypes.REQUEST_QUIZ_FAILED,
};

const requestQuizFailed = (): TypeRequestQuizFailedAction => ({
  type: ActionTypes.REQUEST_QUIZ_FAILED,
});

export type TypeSelectChoiceAction = {
  type: typeof ActionTypes.SELECT_CHOICE,
  payload: {
    selection: TypeSelectionsMap,
  },
};

const selectChoice = (
  selection: TypeSelectionsMap,
): TypeSelectChoiceAction => ({
  type: ActionTypes.SELECT_CHOICE,
  payload: {
    selection,
  },
});

export {
  ActionTypes,
  requestQuiz,
  requestQuizSucceeded,
  requestQuizFailed,
  selectChoice,
};
