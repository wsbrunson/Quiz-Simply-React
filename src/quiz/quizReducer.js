// @flow
import { ActionTypes } from './quizActions';
import type { TypeAction } from '../types/TypeAction.flow';
import type {
  TypeQuestionsMap,
  TypeSelectionsMap,
} from '../types/TypeQuiz.flow';

export type TypeQuizState = {
  questions: TypeQuestionsMap,
  selections: TypeSelectionsMap,
};

const initialState = {
  questions: {},
  selections: {},
};

export default (state: TypeQuizState = initialState, action: TypeAction) => {
  switch (action.type) {
    case ActionTypes.REQUEST_QUIZ_SUCCEEDED:
      return {
        ...state,
        questions: action.payload.questions,
      };
    case ActionTypes.SELECT_CHOICE:
      return {
        ...state,
        selections: {
          ...state.selections,
          ...action.payload.selection,
        },
      };
    default:
      return state;
  }
};
