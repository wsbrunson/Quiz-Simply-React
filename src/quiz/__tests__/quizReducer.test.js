import quizReducer from '../quizReducer';

import * as quizActions from '../quizActions';

describe('Quiz Reducer', () => {
  describe('requestQuizSucceeded', () => {
    it('should add questions to state', () => {
      const state = quizReducer(
        undefined,
        quizActions.requestQuizSucceeded({
          0: {
            id: 0,
          },
        }),
      );

      expect(state).toMatchSnapshot();
    });
  });

  describe('selectChoice', () => {
    it('should add selection', () => {
      const state = quizReducer(
        undefined,
        quizActions.selectChoice({
          '0': '0',
        }),
      );

      expect(state).toMatchSnapshot();
    });

    it('should handle multiple selections', () => {
      const firstSelection = quizReducer(
        undefined,
        quizActions.selectChoice({
          '0': '0',
        }),
      );

      const nextSelection = quizReducer(
        firstSelection,
        quizActions.selectChoice({
          '1': '0',
        }),
      );

      expect(nextSelection).toMatchSnapshot();
    });

    it('should handle multiple selections on the same id', () => {
      const firstSelection = quizReducer(
        undefined,
        quizActions.selectChoice({
          '0': '0',
        }),
      );

      const nextSelection = quizReducer(
        firstSelection,
        quizActions.selectChoice({
          '0': '1',
        }),
      );

      expect(nextSelection).toMatchSnapshot();
    });
  });
});
