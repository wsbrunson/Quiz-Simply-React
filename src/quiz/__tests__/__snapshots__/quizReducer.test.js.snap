// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Reducer requestQuizSucceeded should add questions to state 1`] = `
Object {
  "questions": Object {
    "0": Object {
      "id": 0,
    },
  },
  "selections": Object {},
}
`;

exports[`Quiz Reducer selectChoice should add selection 1`] = `
Object {
  "questions": Object {},
  "selections": Object {
    "0": "0",
  },
}
`;

exports[`Quiz Reducer selectChoice should handle multiple selections 1`] = `
Object {
  "questions": Object {},
  "selections": Object {
    "0": "0",
    "1": "0",
  },
}
`;

exports[`Quiz Reducer selectChoice should handle multiple selections on the same id 1`] = `
Object {
  "questions": Object {},
  "selections": Object {
    "0": "1",
  },
}
`;
